File: 251051/Korneliuk/JPA/src/main/java/com/rest/restapp/service/IssueService.java
package com.rest.restapp.service;

import com.rest.restapp.dto.request.IssueRequestTo;
import com.rest.restapp.dto.response.UserResponseTo;
import com.rest.restapp.dto.response.IssueResponseTo;
import com.rest.restapp.entity.Issue;
import com.rest.restapp.entity.Tag;
import com.rest.restapp.exception.DuplicateException;
import com.rest.restapp.exception.NotFoundException;
import com.rest.restapp.mapper.IssueMapper;
import com.rest.restapp.repository.UserRepository;
import com.rest.restapp.repository.IssueRepository;
import com.rest.restapp.repository.TagRepository;
import jakarta.validation.ValidationException;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
public class IssueService {

    IssueRepository issueRepository;
    UserRepository userRepository;
    TagRepository tagRepository;
    IssueMapper mapper;
    UserService userService;

    @Transactional
    public IssueResponseTo createIssue(IssueRequestTo requestTo) {
        validateIssueRequest(requestTo);
        if (issueRepository.existsByTitle(requestTo.title())) {
            throw new DuplicateException("This issue already exists");
        }

        var user = userRepository
                .findById(requestTo.userId())
                .orElseThrow(() -> new NotFoundException("User with id " + requestTo.userId() + " not found"));

        var issue = mapper.toEntity(requestTo);
        issue.setUser(user);

        if(requestTo.tags() != null) {
            var tags = requestTo.tags().stream()
                    .map(tagName -> tagRepository
                            .findByName(tagName)
                            .orElseGet(() -> tagRepository.save(new Tag(null, tagName, null))))
                    .toList();
            issue.setTags(tags);
        }

        Issue savedIssue = issueRepository.save(issue);
        return mapper.toResponseTo(savedIssue);
    }

    public IssueResponseTo getIssueById(Long id) {
        var issue = issueRepository
                .findById(id)
                .orElseThrow(() -> new NotFoundException("Issue with id " + id + " not found"));
        return mapper.toResponseTo(issue);
    }

    @Transactional(readOnly = true)
    public List<IssueResponseTo> getAllIssues() {
        return issueRepository.findAll().stream()
                .map(mapper::toResponseTo)
                .toList();
    }

    @Transactional
    public IssueResponseTo updateIssue(Long id, IssueRequestTo requestTo) {
        validateIssueRequest(requestTo);
        var existingIssue = issueRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Issue with id " + id + " not found"));

        var user = userRepository.findById(requestTo.userId())
                .orElseThrow(() -> new NotFoundException("User with id " + requestTo.userId() + " not found"));

        mapper.updateEntityFromDto(requestTo, existingIssue);
        existingIssue.setUser(user);
        var updatedIssue = issueRepository.save(existingIssue);
        return mapper.toResponseTo(updatedIssue);
    }

    @Transactional
    public void deleteIssue(Long id) {
        var issue = issueRepository
                .findById(id)
                .orElseThrow(() -> new NotFoundException("Issue with id " + id + " not found"));

        var tags = new ArrayList<>(issue.getTags());

        issue.getTags().clear();
        issueRepository.delete(issue);

        for(Tag tag : tags) {
            tag.getIssues().remove(issue);
            if(tag.getIssues().isEmpty()) {
                tagRepository.delete(tag);
            }
        }
        issueRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public UserResponseTo getUserByIssueId(Long issueId) {
        var issue = issueRepository.findById(issueId)
                .orElseThrow(() -> new NotFoundException("Issue with id " + issueId + " not found"));
        return userService.getUserById(issue.getUser().getId());
    }

    private void validateIssueRequest(IssueRequestTo requestTo) {
        if (requestTo == null) {
            throw new ValidationException("Issue request cannot be null");
        }
        if (requestTo.userId() == null) {
            throw new ValidationException("User ID is required");
        }
        if (requestTo.title() == null || requestTo.title().trim().isEmpty()) {
            throw new ValidationException("Title is required");
        }
        if (requestTo.content() == null || requestTo.content().trim().isEmpty()) {
            throw new ValidationException("Content is required");
        }
    }
}

File: 351002/Mikhalkov/Distcomp.Application/Services/IssueService.cs
using AutoMapper;
using Distcomp.Application.DTOs;
using Distcomp.Application.Exceptions;
using Distcomp.Application.Interfaces;
using Distcomp.Domain.Models;

namespace Distcomp.Application.Services
{
    public class IssueService : IIssueService
    {
        private readonly IRepository<Issue> _issueRepository;
        private readonly IRepository<User> _userRepository; 
        private readonly IMapper _mapper;

        public IssueService(IRepository<Issue> issueRepository, IRepository<User> userRepository, IMapper mapper)
        {
            _issueRepository = issueRepository;
            _userRepository = userRepository;
            _mapper = mapper;
        }

        public IssueResponseTo Create(IssueRequestTo request)
        {
            ValidateIssueRequest(request);

            if (_userRepository.GetById(request.UserId) == null)
                throw new RestException(404, 40401, $"User with id {request.UserId} not found. Cannot create issue.");

            var issue = _mapper.Map<Issue>(request);

            issue.Created = DateTime.UtcNow;
            issue.Modified = DateTime.UtcNow;

            var createdIssue = _issueRepository.Create(issue);
            return _mapper.Map<IssueResponseTo>(createdIssue);
        }

        public IssueResponseTo? GetById(long id)
        {
            var issue = _issueRepository.GetById(id);
            if (issue == null)
                throw new RestException(404, 40402, $"Issue with id {id} not found");

            return _mapper.Map<IssueResponseTo>(issue);
        }

        public IEnumerable<IssueResponseTo> GetAll()
        {
            return _mapper.Map<IEnumerable<IssueResponseTo>>(_issueRepository.GetAll());
        }

        public IssueResponseTo Update(long id, IssueRequestTo request)
        {
            var existingIssue = _issueRepository.GetById(id);
            if (existingIssue == null)
                throw new RestException(404, 40402, $"Cannot update: Issue with id {id} not found");

            ValidateIssueRequest(request);

            if (existingIssue.UserId != request.UserId && _userRepository.GetById(request.UserId) == null)
                throw new RestException(404, 40401, $"Cannot update issue: New User with id {request.UserId} not found");

            _mapper.Map(request, existingIssue);
            existingIssue.Id = id;
            existingIssue.Modified = DateTime.UtcNow;

            _issueRepository.Update(existingIssue);

            return _mapper.Map<IssueResponseTo>(existingIssue);
        }

        public bool Delete(long id)
        {
            var existingIssue = _issueRepository.GetById(id);
            if (existingIssue == null)
                throw new RestException(404, 40402, $"Cannot delete: Issue with id {id} not found");

            return _issueRepository.Delete(id);
        }

        private void ValidateIssueRequest(IssueRequestTo request)
        {
            if (request.Title.Length < 2 || request.Title.Length > 64)
                throw new RestException(400, 40005, "Title length must be between 2 and 64");

            if (request.Content.Length < 4 || request.Content.Length > 2048)
                throw new RestException(400, 40006, "Content length must be between 4 and 2048");
        }
    }
}